Lib/threading.py

```diff
@@ -775,8 +775,11 @@ class BrokenBarrierError(RuntimeError):
 def _newname(name_template):
     return name_template % _counter()
 
-# Active thread administration
-_active_limbo_lock = _allocate_lock()
+# Active thread administration.
+#
+# bpo-44422: Use a reentrant lock to allocate reentrant calls to functions like
+# threading.enumerate().
+_active_limbo_lock = RLock()
```

Comment:
The *RLock()* is the right way to go to prevent self-deadlock.  

We could temporarily turn GC off but that seems hard to do in a way that reliably restores the prior enabled/disabled mode.

Reply:
> We could temporarily turn GC off but that seems hard to do in a way that reliably restores the prior enabled/disabled mode.

Sure, that's the other option that I considered. But gc.disable() is process-wide: it affects all Python threads, and so it might have surprising side effects. Some code might rely on the current exact GC behavior.

Another option would be to rewrite the code in C. The problem is that other functions rely on this lock, like active_count(). I would prefer to not have to rewrite "half" of threading.py in C. Using a RLock is less intrusive.